### PR TITLE
Prevent double bind on same address.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #1123

1. Solves the issue where we allowed binding the same `{address:port}` twice;
2. Adds a new CI run (build issue1123 integration test);

Doesn't close the issue yet though, as we still need to "fake" the ip as the impersonated pod ip, which involves some changes to `getsockname`, `bind`, and probably `listen`.

### CODE REVIEW

File: mirrord/layer/src/socket/ops.rs [hunk 1]
```diff
@@ -165,6 +165,21 @@ pub(super) fn bind(
...
 
+    // Check if the user's requested address isn't already in use, even though it's not actually
+    // bound, as we bind to a different address, but if we don't check for this then we're
+    // changing normal socket behavior (see issue #1123).
+    if SOCKETS.iter().any(|socket| match &socket.state {
+        SocketState::Initialized => false,
+        SocketState::Bound(bound) | SocketState::Listening(bound) => {
+            bound.requested_address == requested_address
+        }
+        SocketState::Connected(connected) => SocketAddr::try_from(connected.local_address.clone())
```
Comment (aviramha) [inline mirrord/layer/src/socket/ops.rs:176]: This shouldn't return bound AFAIK. Once connection is made a new socket is created and bound/listening socket state stays the same and can also be closed letting other binding/listening happen

Comment (meowjesty) [reply mirrord/layer/src/socket/ops.rs:176]: You mean for the `Socket::Connected` state only?

File: mirrord/layer/tests/common/mod.rs [hunk 2]
```diff
@@ -124,6 +124,7 @@ impl TestProcess {
     }
 
     pub fn assert_stdout_contains(&self, string: &str) {
+        println!("{:#?}", self.stdout.lock().unwrap().as_str());
```
Comment (Razz4780) [inline mirrord/layer/tests/common/mod.rs:127]: Is this needed? Will clutter the output for all tests

File: mirrord/layer/tests/issue1054.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    println!("Application subscribed to port, sending tcp messages.");
+
+    let new_connection_id = layer_connection
+        .send_new_connection(application.get_app_port())
+        .await;
+    layer_connection.send_close(new_connection_id).await;
+
+    test_process.wait().await;
+    test_process.assert_stdout_contains("test issue 1054: START");
+    test_process.assert_stdout_contains("test issue 1054: SUCCESS");
```
Comment (Razz4780) [inline mirrord/layer/tests/issue1054.rs:44]: Nitpick: we could just check the return code

File: mirrord/layer/tests/issue1123.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            ],
+            None,
+        )
+        .await;
+
+    println!("Application subscribed to port, sending tcp messages.");
+
+    test_process.wait().await;
+    test_process.assert_stdout_contains("test issue 1123: START");
+    test_process.assert_stdout_contains("test issue 1123: SUCCESS");
```
Comment (Razz4780) [inline mirrord/layer/tests/issue1123.rs:36]: Nitpick: we could just check the return code

**Review by Razz4780 — CHANGES_REQUESTED** [review]

**Review by Razz4780 — APPROVED** [review]
Looks good!

**Pull request merged**